### ISSUE TITLE
We have two hex! macros

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#1987]
We use two macros `hex!`, one is `hex_conservative::test_hex_unwrap` and uses `FromHex` and returns a `Vec` (ie, allocates memory on the heap). The other comes is `hex_lit::hex`.

One obvious difference that is not easily overcome is `hex_conservative::test_hex_unwrap` can take an expression and not just a const. Apart from that it is not obvious to me when to use which

- `hex_lit::hex` uses the stack so when is a hex string so big that the heap is better?
- In test code do allocations matter?

The **real** question is, do we need both?

**Comment by apoelstra** [comment]
I think `hex_lit` can probably just be folded into `hex-conservative` now that we have it.

The history here is that `hex_lit` was introduced a year ago in https://github.com/rust-bitcoin/rust-bitcoin/pull/1190 based on disussion in #1189. Meanwhile hex-conservative did not have a release for 9 more months.

**Comment by elichai** [comment]
Also, unless I'm misremembering it's used only it tests, so I'm not sure if the runtime perf improvements are bigger than the bigger compile time run. (as in tests you usually care about the sum of those)

**Comment by tcharding** [comment]
The `hex_lit` is used at least once outside of test code, in `blockdata::constants` but it looks to me like an allocation there would be fine.

**Comment by apoelstra** [comment]
If it's really only once we could just remove the usage and use verbose `[0xab, 0xcd]` syntax.

**Comment by tcharding** [comment]
Will investigate further.

**Comment by vincenzopalazzo** [comment]
Do you think that it is better also to expose one of these macros?

It is very useful for some applications that work with two different versions of rust-bitcoin.

**Comment by tcharding** [comment]
The heap allocation one is exposed already from `hex_conservative`: https://docs.rs/hex-conservative/latest/hex_conservative/macro.test_hex_unwrap.html

**Comment by tcharding** [comment]
> I think `hex_lit` can probably just be folded into `hex-conservative` now that we have it.

Do you agree with this statement @Kixunil? If so we can raise an issue on the `hex-conservative` crate and just close this one I rekon.

**Comment by Kixunil** [comment]
Yes, I do, I already said it elsewhere.

I now also noticed this claim:

> unless I'm misremembering it's used only it tests

There's at least one non-test case - when constructing genesis blocks.

I think many cases that currently can't use the `const` version could be easily rewritten to support it. I remember already doing this somewhere. However it is true that it's not 100% of them. I think we could jut not use a macro there.

**Comment by tcharding** [comment]
Closing, between the PR linked above and #5645 this is done.
